elements/lisk-tree/src/merkle_tree/utils.ts

```diff
@@ -193,3 +201,136 @@ export const calculateMerkleRoot = ({ value, appendPath, size }: MerkleRootInfo)
+	const rightHash = type === NodeType.BRANCH ? value.slice(-1 * NODE_HASH_SIZE) : Buffer.alloc(0);
+	const leftHash =
+		type === NodeType.BRANCH
+			? value.slice(-2 * NODE_HASH_SIZE, -1 * NODE_HASH_SIZE)
+			: Buffer.alloc(0);
+
+	return {
+		type,
+		hash: nodeHash,
+		value,
+		layerIndex,
+		nodeIndex,
+		rightHash,
+		leftHash,
+	};
+};
+
+export const buildLeaf = (value: Buffer, nodeIndex: number, preHashedLeaf?: boolean) => {
+	const nodeIndexBuffer = Buffer.alloc(NODE_INDEX_SIZE);
+	nodeIndexBuffer.writeInt32BE(nodeIndex, 0);
+	// As per protocol nodeIndex is not included in hash
+	const leafValueWithoutNodeIndex = Buffer.concat(
+		[LEAF_PREFIX, value],
+		LEAF_PREFIX.length + value.length,
+	);
+	const leafHash = preHashedLeaf ? value : hash(leafValueWithoutNodeIndex);
+	// We include nodeIndex into the value to allow for nodeIndex retrieval for leaf nodes
+	const leafValueWithNodeIndex = Buffer.concat(
+		[LEAF_PREFIX, nodeIndexBuffer, value],
+		LEAF_PREFIX.length + nodeIndexBuffer.length + value.length,
+	);
+
+	return {
+		leafValueWithNodeIndex,
+		leafHash,
+	};
+};
+
+export const buildBranch = (
+	leftHashBuffer: Buffer,
+	rightHashBuffer: Buffer,
+	layerIndex: number,
+	nodeIndex: number,
+) => {
+	const layerIndexBuffer = Buffer.alloc(LAYER_INDEX_SIZE);
+	const nodeIndexBuffer = Buffer.alloc(NODE_INDEX_SIZE);
+	layerIndexBuffer.writeInt8(layerIndex, 0);
+	nodeIndexBuffer.writeInt32BE(nodeIndex, 0);
+
+	const branchValue = Buffer.concat(
+		[BRANCH_PREFIX, layerIndexBuffer, nodeIndexBuffer, leftHashBuffer, rightHashBuffer],
+		BRANCH_PREFIX.length +
+			layerIndexBuffer.length +
+			nodeIndexBuffer.length +
+			leftHashBuffer.length +
+			rightHashBuffer.length,
+	);
+	const branchHash = generateHash(BRANCH_PREFIX, leftHashBuffer, rightHashBuffer);
+
+	return {
+		branchHash,
+		branchValue,
+	};
+};
+
+export const createNewBranchNode = (
+	location: NodeLocation,
+	pairHash: Buffer,
+	currentHash: Buffer,
+	pairSide: NodeSide,
+) => {
+	const leftHashBuffer = pairSide === NodeSide.LEFT ? pairHash : currentHash;
+	const rightHashBuffer = pairSide === NodeSide.RIGHT ? pairHash : currentHash;
+	const newNodeData = buildBranch(
+		leftHashBuffer,
+		rightHashBuffer,
+		location.layerIndex,
+		location.nodeIndex,
+	);
+	const newNode = generateNode(newNodeData.branchHash, newNodeData.branchValue);
+
+	return newNode;
+};
+
+export const createNewLeafNode = (value: Buffer, nodeIndex: number, preHashed: boolean) => {
+	const newNodeData = buildLeaf(value, nodeIndex, preHashed);
+	const newNode = generateNode(newNodeData.leafHash, newNodeData.leafValueWithNodeIndex);
+
+	return newNode;
+};
+
+export const getLocationFromIndex = (index: number, size: number) => {
+	const treeHeight = Math.ceil(Math.log2(size)) + 1;
+
+	const serializedIndexBinaryString = index.toString(2);
+	const indexBinaryString = serializedIndexBinaryString.substring(
+		1,
+		serializedIndexBinaryString.length,
+	);
+	const location = {
```

Comment:
I think at some point here we should check that the index is not larger than the layer size